index.js

```diff
@@ -458,50 +487,57 @@ XiaomiRoborockVacuum.prototype = {
-            that.cleaning = false
-            that.lastrobotcleaning = that.cleaning;
-
-            // Cleaning stoped => pausing not possible
-            if (that.pause) {
-                that.pausepossible = false
-                that.lastrobotpausecleaning = that.pausepossible;
-                that.pauseService.getCharacteristic(Characteristic.On).updateValue(that.pausepossible);
+                return that.device.activateCleaning()
+                    // IS THIS REALLY NEEDED?
+                    // .then(() => {
+                        // that.cleaning = true
+                        // that.lastrobotcleaning = that.cleaning;
+                        // that.charging = false;
+                        // that.lastrobotcharging = that.charging;
+                        // that.batteryService.getCharacteristic(Characteristic.ChargingState).updateValue(Characteristic.ChargingState.NOT_CHARGING);
+
+                        // Cleaning => leaves dock
+                        // if (that.dock) {
+                            // that.docked = false;
+                            // that.dockService.getCharacteristic(Characteristic.OccupancyDetected).updateValue(that.docked);
+                            // log.info('ACT setCleaning - dock | ' + that.model + ' | Docked set to %s.', that.docked);
+                        // }
+
+                        // Cleaning => pausing possible
+                        // if (that.pause) {
+                        //     that.pausepossible = true;
+                        //     that.lastrobotpausecleaning = that.pausepossible;
+                        //     that.pauseService.getCharacteristic(Characteristic.On).updateValue(that.pausepossible);
+                        //     log.info('ACT setCleaning - pause | ' + that.model + ' | PausePossible set to %s.', that.pausepossible);
+                        // }
+                    // });
+            } else if (!state) { // Stop cleaning
+                log.info('ACT setCleaning | ' + that.model + ' | Stop cleaning and go to charge.');
+                return that.device.activateCharging() // Charging works for 1st, not for 2nd
+                // IS THIS REALLY NEEDED?
+                // .then(() => {
+                    // that.cleaning = false
+                    // that.lastrobotcleaning = that.cleaning;
+
+                    // // Cleaning stoped => pausing not possible
+                    // if (that.pause) {
+                    //     that.pausepossible = false
+                    //     that.lastrobotpausecleaning = that.pausepossible;
+                    //     that.pauseService.getCharacteristic(Characteristic.On).updateValue(that.pausepossible);
+                    // }
+                // });
             }
-
-        }
-        callback();
+        })
+        .then(() => callback())
+        .catch((err) => {
+            // SHOULD WE RECONNECT AND RETRY?
```

Comment:
That would be good - every 120 seconds? 😄

Reply:
Deal! I didn't implement it because I wanted to see how the miio interface behaves. If it autoreconnects on every new command or so...
I'll check their implementation if there's anything already in place :)

Reply:
So far, since Saturday, the automation I’ve got in place has not failed and the cleaning status is correctly reported (like in #36, I was sometimes finding the automation “started” the robot but it actually was not moving and, since nothing actually changed for the robot, it was never emitting the “finished cleaning” event so, according to HomeKit, it was still on).
I’ll keep an eye on it for the rest of the week since I cannot jump on this until the weekend.
But maybe there is some sort of automatic reconnection when triggering any of the methods in the miio API.
Like I said, I’ll double check this weekend, but keeping an eye on the behavior during the week